File: app/src/main/java/jp/juggler/subwaytooter/view/ColumnStripLinearLayout.java
package jp.juggler.subwaytooter.view;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Rect;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.view.View;
import android.widget.LinearLayout;

import jp.juggler.subwaytooter.R;
import jp.juggler.subwaytooter.Styler;

public class ColumnStripLinearLayout extends LinearLayout {
	public ColumnStripLinearLayout( Context context ){
		super( context );
		init();
	}
	
	public ColumnStripLinearLayout( Context context, @Nullable AttributeSet attrs ){
		super( context, attrs );
		init();
	}
	
	public ColumnStripLinearLayout( Context context, @Nullable AttributeSet attrs, int defStyleAttr ){
		super( context, attrs, defStyleAttr );
		init();
	}
	
	final Paint paint = new Paint();
	final Rect rect = new Rect();
	int h ;
	void init(){
		h = (int)(0.5f + 2f * getResources().getDisplayMetrics().density );
	}
	
	int first;
	int last;
	float slide_ratio;
	
	public void setColumnRange(int first,int last,float slide_ratio){
		if( this.first == first && this.last == last && this.slide_ratio == slide_ratio ) return;
		this.first = first;
		this.last = last;
		this.slide_ratio = slide_ratio;
		invalidate();
	}
	
	int color;
	public void setColor( int color ){
		if( color == 0 ) color =  Styler.getAttributeColor(getContext(), R.attr.colorAccent);
		if( this.color == color ) return;
		this.color = color;
		invalidate();
	}
	
	
	@Override protected void dispatchDraw(Canvas canvas) {
		super.dispatchDraw(canvas);
		
		if( first < 0 || last >= getChildCount() ) return;
		
		View child = getChildAt( first );
		rect.left = child.getLeft();
		child = getChildAt( last );
		rect.right = child.getRight();
		
		if( slide_ratio != 0f){
			child = getChildAt( first );
			int w = child.getWidth();
			int slide = (int)(0.5f + slide_ratio * w);
			rect.left += slide;
			rect.right += slide;
		}
		
		rect.top = 0;
		rect.bottom = h;
		paint.setColor( color );
		canvas.drawRect( rect,paint );
	}
	
}

File: app/src/main/java/jp/juggler/subwaytooter/view/ColumnStripLinearLayout.kt
package jp.juggler.subwaytooter.view

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Rect
import android.util.AttributeSet
import android.widget.LinearLayout

import jp.juggler.subwaytooter.R
import jp.juggler.subwaytooter.Styler

// 画面下のカラムストリップのLinearLayout
// 可視範囲を示すインジケーターを表示する
class ColumnStripLinearLayout : LinearLayout {
	
	// 可視範囲
	private var visibleFirst : Int = 0
	private var visibleLast : Int = 0
	private var visibleSlideRatio : Float = 0.toFloat()
	
	// インジケーターの高さ
	private val indicatorHeight : Int
	
	// インジケーターの色
	var indicatorColor : Int = 0
		set(value) {
			val c = if(value != 0) value else Styler.getAttributeColor(context, R.attr.colorAccent)
			if(field != c) {
				field = c
				invalidate()
			}
		}
	
	init {
		indicatorHeight = (0.5f + 2f * resources.displayMetrics.density).toInt()
	}
	
	constructor(context : Context) : super(context)
	constructor(context : Context, attrs : AttributeSet?) : super(context, attrs)
	constructor(context : Context, attrs : AttributeSet?, defStyleAttr : Int) : super(context, attrs, defStyleAttr)
	
	fun setVisibleRange(first : Int, last : Int, slide_ratio : Float) {
		if(this.visibleFirst == first
			&& this.visibleLast == last
			&& this.visibleSlideRatio == slide_ratio
			) return
		this.visibleFirst = first
		this.visibleLast = last
		this.visibleSlideRatio = slide_ratio
		invalidate()
	}
	
	private val paint = Paint()
	private val rect = Rect()
	
	override fun dispatchDraw(canvas : Canvas) {
		super.dispatchDraw(canvas)
		
		if(visibleFirst < 0 || visibleLast >= childCount) return
		
		val childFirst = getChildAt(visibleFirst)
		val childLast = getChildAt(visibleLast)
		
		val x = if(visibleSlideRatio != 0f) {
			(0.5f + visibleSlideRatio * childFirst.width).toInt()
		}else{
			0
		}

		rect.left = childFirst.left +x
		rect.right = childLast.right+x
		rect.top = 0
		rect.bottom = indicatorHeight

		paint.color = indicatorColor

		canvas.drawRect(rect, paint)
	}
	
}
